core/crypto/abac/attributes.go

```diff
@@ -0,0 +1,267 @@
+	if !strings.HasPrefix(header, headerPrefix) {
+		return nil, errors.New("Invalid header")
+	}
+	headerBody := strings.Replace(header, headerPrefix, "", 1)
+	tokens := strings.Split(headerBody, "#")
+	result := make(map[string]int)
+
+	for _, token := range tokens {
+		pair := strings.Split(token, "->")
+
+		if len(pair) == 2 {
+			key := pair[0]
+			valueStr := pair[1]
+			value, err := strconv.Atoi(valueStr)
+			if err != nil {
+				return nil, err
+			}
+			result[key] = value
+		}
+	}
+
+	return result, nil
+}
+
+//ReadAttributeHeader read the header of the attributes.
+func ReadAttributeHeader(tcert *x509.Certificate, headerKey []byte) (map[string]int, bool, error) {
+	var err error
+	var headerRaw []byte
+	encrypted := false
+	if headerRaw, err = utils.GetCriticalExtension(tcert, TCertAttributesHeaders); err != nil {
+		return nil, encrypted, err
+	}
+	headerStr := string(headerRaw)
+	var header map[string]int
+	header, err = ParseAttributesHeader(headerStr)
+	if err != nil {
+		if headerKey == nil {
+			return nil, false, errors.New("Is not possible read an attribute encrypted without the headerKey")
+		}
+		headerRaw, err = DecryptAttributeValue(headerKey, headerRaw)
+
+		if err != nil {
+			return nil, encrypted, errors.New("error decrypting header value '" + err.Error() + "''")
+		}
+		headerStr = string(headerRaw)
+		header, err = ParseAttributesHeader(headerStr)
+		if err != nil {
+			return nil, encrypted, err
+		}
+		encrypted = true
+	}
+	return header, encrypted, nil
+}
+
+//ReadTCertAttributeByPosition read the attribute stored in the position "position" of the tcert.
+func ReadTCertAttributeByPosition(tcert *x509.Certificate, position int) ([]byte, error) {
+	if position < 0 {
+		return nil, fmt.Errorf("Invalid attribute position. Received [%v]", position)
+	}
+
+	oid := asn1.ObjectIdentifier{1, 2, 3, 4, 5, 6, 9 + position}
+	value, err := utils.GetCriticalExtension(tcert, oid)
+	if err != nil {
+		return nil, err
+	}
+	return value, nil
+}
+
+//ReadTCertAttribute read the attribute with name "attributeName" and returns the value and a boolean indicating if the returned value is encrypted or not.
+func ReadTCertAttribute(tcert *x509.Certificate, attributeName string, headerKey []byte) ([]byte, bool, error) {
+	header, encrypted, err := ReadAttributeHeader(tcert, headerKey)
+	if err != nil {
+		return nil, false, err
+	}
+	position := header[attributeName]
+	if position == 0 {
+		return nil, encrypted, errors.New("Failed attribute '" + attributeName + "' doesn't exists in the TCert.")
+	}
+	value, err := ReadTCertAttributeByPosition(tcert, position)
+	if err != nil {
+		return nil, encrypted, err
+	}
+	return value, encrypted, nil
+}
+
+//EncryptAttributeValue encrypts "attributeValue" using "attributeKey"
+func EncryptAttributeValue(attributeKey []byte, attributeValue []byte) ([]byte, error) {
+	value := append(attributeValue, padding...)
+	return primitives.CBCPKCS7Encrypt(attributeKey, value)
```

Comment:
Because the attributes are used in multiple places, shall we use GCM mode to protect against malleability attacks? The primitives package offers such an encryption mode.